### PR TITLE
[SPARK-22908][SS] Roll forward continuous processing Kafka support with fix to continuous Kafka data reader

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

The Kafka reader is now interruptible and can close itself.
## How was this patch tested?

I locally ran one of the ContinuousKafkaSourceSuite tests in a tight loop. Before the fix, my machine ran out of open file descriptors a few iterations in; now it works fine.

### CODE REVIEW

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 1]
```diff
@@ -192,11 +193,18 @@ class KafkaContinuousDataReader(
...
-        failOnDataLoss)
+      // Our consumer.get is not interruptible, so we have to set a low poll timeout, leaving
+      // interrupt points to end the query rather than waiting for new data that might never come.
+      // We simply swallow any resulting timeout exceptions and retry.
+      try {
+        r = consumer.get(
+          nextKafkaOffset,
+          untilOffset = Long.MaxValue,
+          pollTimeoutMs = 1000,
+          failOnDataLoss = false)
```
Comment (mgaido91) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: may I kindly ask you why we have to change this?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 2]
```diff
@@ -192,11 +193,18 @@ class KafkaContinuousDataReader(
...
+      } catch {
+        case _: TimeoutException => ()
```
Comment (mgaido91) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: nit: I think this can be `case _: TimeoutException =>` or even better we might add a comment like `case _: TimeoutException => // ignore the exception and continue polling` or what you think is more appropriate

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala (hunk 1 above)
Comment (jose-torres) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: Kafka consumers don't respect thread interrupts, and our consumer wrapper only partly mitigates this by running them in an UninterruptibleThread. As it stands, even when its parent task ends, the reader will never actually close until it reads a new value or the JVM dies.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala (hunk 2 above)
Comment (jose-torres) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: Done.

**Comment by jose-torres** [comment]
The base PR was reverted, so I'm going to roll it back into here soon.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 3]
```diff
@@ -192,11 +194,26 @@ class KafkaContinuousDataReader(
...
+      } catch {
+        // We didn't read within the timeout. We're supposed to block indefinitely for new data, so
+        // swallow and ignore this.
+        case _: TimeoutException =>
+        // Data loss is reported on both sides - but we expect scenarios where the offset we're
+        // looking for isn't available yet. Don't propagate the exception unless it's because
+        // the offset we're looking for is below the available range.
+        case e: IllegalStateException
+          if e.getCause.isInstanceOf[OffsetOutOfRangeException] &&
+            !(consumer.getAvailableOffsetRange().earliest > nextKafkaOffset) =>
```
Comment (zsxwing) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: I would change this condition to `this.getAvailableOffsetRange().latest == nextKafkaOffset` for safety.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 4]
```diff
@@ -192,11 +194,30 @@ class KafkaContinuousDataReader(
...
+          failOnDataLoss)
+      } catch {
+        // We didn't read within the timeout. We're supposed to block indefinitely for new data, so
+        // swallow and ignore this.
+        case _: TimeoutException =>
+        // This is a failOnDataLoss exception. Retry if nextKafkaOffset is within the data range,
+        // or if it's the endpoint of the data range (i.e. the "true" next offset).
+        case e: IllegalStateException =>
+          val range = consumer.getAvailableOffsetRange()
+          if (e.getCause.isInstanceOf[OffsetOutOfRangeException] &&
```
Comment (zsxwing) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: nit: move this condition above like `case e: IllegalStateException  if e.getCause.isInstanceOf[OffsetOutOfRangeException] =>`

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        r = consumer.get(
+          nextKafkaOffset,
+          untilOffset = Long.MaxValue,
+          pollTimeoutMs = 1000,
+          failOnDataLoss)
+      } catch {
+        // We didn't read within the timeout. We're supposed to block indefinitely for new data, so
+        // swallow and ignore this.
+        case _: TimeoutException =>
+        // This is a failOnDataLoss exception. Retry if nextKafkaOffset is within the data range,
```
Comment (tdas) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: add an empty line to distinguish that comment below this belongs to next case

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    var r: ConsumerRecord[Array[Byte], Array[Byte]] = null
+    while (r == null) {
+      if (TaskContext.get().isInterrupted() || TaskContext.get().isCompleted()) return false
+      // Our consumer.get is not interruptible, so we have to set a low poll timeout, leaving
+      // interrupt points to end the query rather than waiting for new data that might never come.
+      try {
+        r = consumer.get(
+          nextKafkaOffset,
+          untilOffset = Long.MaxValue,
+          pollTimeoutMs = 1000,
```
Comment (tdas) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: Make this configurable. In fact, we already have kafka option "kafkaConsumer.pollTimeoutMs" which means roughly this (search for it here https://spark.apache.org/docs/latest/structured-streaming-kafka-integration.html). It has a low default value of 512ms. This should be good.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        case e: IllegalStateException =>
+          val range = consumer.getAvailableOffsetRange()
+          if (e.getCause.isInstanceOf[OffsetOutOfRangeException] &&
+              range.latest >= nextKafkaOffset && range.earliest <= nextKafkaOffset) {
+            // retry
+          } else {
+            throw e
+          }
+      }
+    }
```
Comment (tdas) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: Since we ran into this issue, we should add a test that ensures that all the KafkaContinuous tasks eventually terminate after every unit test. This could be added in `afterEach`. You could use a listener to track num tasks started and terminated and ensure that it becomes 0 after every test.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousDataSourceRDDIter.scala [hunk 8]
```diff
@@ -177,6 +176,7 @@ class DataReaderThread(
   private[continuous] var failureReason: Throwable = _
 
   override def run(): Unit = {
+    TaskContext.setTaskContext(context)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousDataSourceRDDIter.scala]: This line was added since the last PR. Why is this needed?

**Comment by tdas** [comment]
Why dont we have full failOnDataLoss support any more with this change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousDataSourceRDDIter.scala (hunk 8 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousDataSourceRDDIter.scala]: TaskContext is a ThreadLocal. We need to get it in the Kafka data reader to check if an interrupt has come.

I argue that this doesn't represent an abstraction failure in the V2 API, but a workaround to the fact that Kafka consumers are badly behaved and do not respect interrupts.

**Comment by jose-torres** [comment]
That was an old part of the description, and no longer applies to the current change.

We have almost full failOnDataLoss support, with the exception that searching for an offset greater than the available range is not a failure. This is a data loss scenario in the microbatch case, because when it happens it means that an offset within the preconstructed batch range was no longer available. I think that this is not a data loss scenario in the continuous case, since:
* we expect to block forever until the next row shows up
* we do not have a secondary consumer that will also expect to see this offset

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala (hunk 7 above)
Comment (jose-torres) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaContinuousReader.scala]: Done. In a future PR (I want to minimize the delta on a roll-forward), I'm going to figure out how to get this in ContinuousTest for the benefit of future sources.

**Comment by jose-torres** [comment]
The builds at commit 3fe76e3 were on an incomplete version of the PR, so their failures are expected.

Of the 6 builds at target commit 0efc8c5, 5 passed and 1 failed. The failure was due to a stream.processAllAvailable() call in the Kafka suite; we already knew this method is inherently flake-prone for continuous processing, and had attempted to remove usages of it. I'm going to do another pass to get the rest, and then kick off 5 more attempts.

**Comment by jose-torres** [comment]
In commit 4bb9c3f, there are no occurrences remaining of the string "processAllAvailable" in KafkaContinuousSourceSuite.scala, KafkaContinuousSinkSuite.scala, KafkaContinuousTest.scala, or StreamTest.scala.

There are four occurrences in KafkaSourceSuite.scala, two within the MicroBatch suite and two matched to not happen in ContinuousExecution. (One test with a foreach sink was moved to the MicroBatch suite, because it was executing in microbatch mode anyway since we haven't updated foreach for continuous processing.)

I believe this is exhaustive.

**Comment by jose-torres** [comment]
Kicked off the 5 runs.

I noticed one of the earlier runs actually failed in PySpark. I don't know of a plausible mechanism by which this PR could cause that.

**Comment by jose-torres** [comment]
All tests seem to have gotten a kill -9 at the same clock time (1:05), despite being on different machines and at different points in the run. Poking around the build history, this same thing seems to have happened to other runs active at 1:05.

I'm calling shenanigans and starting a few more.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        query.lastExecution.logical.collectFirst {
+          case DataSourceV2Relation(_, r: KafkaContinuousReader) => r
+        }.exists(_.knownPartitions.size == newCount),
+        s"query never reconfigured to $newCount partitions")
+    }
+  }
+
+  // Continuous processing tasks end asynchronously, so test that they actually end.
+  private val tasksEndedListener = new SparkListener() {
+    val activeTaskIds = mutable.Set[Long]()
```
Comment (tdas) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala]: This is not a synchronized set. We are accessing this from multiple threads (listener thread is adding/removing, and test thread is checking if its empty). So this is an incorrect test. 
In fact, this can be made simpler using just an AtomicInteger.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+    override def onTaskStart(start: SparkListenerTaskStart): Unit = {
+      activeTaskIds.add(start.taskInfo.taskId)
+    }
+
+    override def onTaskEnd(end: SparkListenerTaskEnd): Unit = {
+      activeTaskIds.remove(end.taskInfo.taskId)
+    }
+  }
+  override def beforeEach(): Unit = {
```
Comment (tdas) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala]: add an empty line.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    spark.sparkContext.addSparkListener(tasksEndedListener)
```
Comment (tdas) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala]: you are not calling super.beforeEach. This may have unforeseen circumstances.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  override def afterEach(): Unit = {
+    eventually(timeout(streamingTimeout)) {
+      assert(tasksEndedListener.activeTaskIds.isEmpty)
+    }
+    spark.sparkContext.removeSparkListener(tasksEndedListener)
```
Comment (tdas) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaContinuousTest.scala]: call super.afterEach()

**Comment by tdas** [comment]
LGTM, assuming tests pass.

**Comment by tdas** [comment]
Great. Merging this to master and 2.3.
